Sources/Sentry/SentryTracerConcurrency.mm

```diff
@@ -1,49 +1,123 @@
+    NSMutableSet<SentryTracer *> *>
+    SentryProfilerTracerMapping;
+static SentryProfilerTracerMapping *_gProfilersToTracers;
+
+// provided for fast access to a profiler given a tracer
+typedef NSMutableDictionary</* SentryTracer.tracerId */ NSString *, SentryProfiler *>
+    SentryTracerProfilerMapping;
+static SentryTracerProfilerMapping *_gTracersToProfilers;
+
 std::mutex _gStateLock;
 
 void
-trackTracerWithID(SentryId *traceID)
+trackProfilerForTracer(SentryProfiler *profiler, SentryTracer *tracer)
 {
     std::lock_guard<std::mutex> l(_gStateLock);
 
-    if (_gInFlightTraceIDs == nil) {
-        _gInFlightTraceIDs = [NSMutableSet<NSString *> set];
+    const auto profilerKey = profiler.profileId.sentryIdString;
+    const auto tracerKey = tracer.traceId.sentryIdString;
+
+    SENTRY_LOG_DEBUG(
+        @"Tracking relationship between profiler id %@ and tracer id %@", profilerKey, tracerKey);
+
+    NSCAssert((_gProfilersToTracers == nil && _gTracersToProfilers == nil)
+            || (_gProfilersToTracers != nil && _gTracersToProfilers != nil),
+        @"Both structures must be initialized simultaneously.");
+
+    if (_gProfilersToTracers == nil) {
+        _gProfilersToTracers =
+            [SentryProfilerTracerMapping dictionaryWithObject:[NSMutableSet setWithObject:tracer]
+                                                       forKey:profilerKey];
+        _gTracersToProfilers = [SentryTracerProfilerMapping dictionaryWithObject:profiler
+                                                                          forKey:tracerKey];
+        return;
     }
-    const auto idString = traceID.sentryIdString;
-    SENTRY_LOG_DEBUG(@"Adding tracer id %@", idString);
-    [_gInFlightTraceIDs addObject:idString];
+
+    if (_gProfilersToTracers[profilerKey] == nil) {
+        _gProfilersToTracers[profilerKey] = [NSMutableSet setWithObject:tracer];
+    } else {
+        [_gProfilersToTracers[profilerKey] addObject:tracer];
+    }
+
+    _gTracersToProfilers[tracerKey] = profiler;
 }
 
-void
-stopTrackingTracerWithID(SentryId *traceID, SentryConcurrentTransactionCleanupBlock cleanup)
+SentryProfiler *_Nullable profilerForTracer(SentryTracer *tracer)
 {
     std::lock_guard<std::mutex> l(_gStateLock);
+    NSCAssert(_gTracersToProfilers != nil,
+        @"Structure should have already been initialized by the time it is being queried");
+    return _gTracersToProfilers[tracer.traceId.sentryIdString];
+}
 
-    const auto idString = traceID.sentryIdString;
-    SENTRY_LOG_DEBUG(@"Removing trace id %@", idString);
-    [_gInFlightTraceIDs removeObject:idString];
-    if (_gInFlightTraceIDs.count == 0) {
-        SENTRY_LOG_DEBUG(@"Last in flight tracer completed, performing cleanup.");
-        cleanup();
-    } else {
-        SENTRY_LOG_DEBUG(@"Waiting on %lu other tracers to complete: %@.", _gInFlightTraceIDs.count,
-            _gInFlightTraceIDs);
+SentryProfiler *_Nullable profilerForFinishedTracer(SentryTracer *tracer)
```

Comment:
`h`: This method only gets called in `SentryProfiler.createProfilingEnvelopeItemForTransaction`, which is called by, `SentryTracer.captureTransactionWithProfile`, which is called by `SentryTracer.finishInternal`. There is no guarantee that `SentryTracer.finishInternal` will actually call `SentryTracer.captureTransactionWithProfile`. Therefore, it can happen that `_gProfilersToTracers` and `_gTracersToProfilers` keep references to `SentryTracer` and `SentryProfile` which should have been deallocated. This can lead to infinite memory growth, or am I missing something? We could solve this by keeping weak references here or we need some other way of cleaning up the two dictionaries. I think this problem already existed before, so we can also address in an extra PR.

Reply:
We can't use weak refs because there won't be anything else keeping valid profilers alive. The simplest solution I can think of is to call some cleanup function that removes them from the dicts at every early return in the codepath that would otherwise lead to capturing a transaction. 

Are there any early returns in `finishInternal` that _shouldn't_ do such cleanup, because it will be called again later?

And conversely, are there any callers to `finishInternal` that have early returns where we'd also need to do this cleanup because there won't be another call to `finishInternal`?

Happy to tackle this in a separate PR as stated.

Reply:
I did this in two change: https://github.com/getsentry/sentry-cocoa/pull/3154 to clean up profilers for discarded transactions, and weak references in https://github.com/getsentry/sentry-cocoa/pull/3155